src/kaleidoscope/MatrixAddr.h

```diff
@@ -0,0 +1,248 @@
+  //
+  constexpr MatrixAddr(const ThisType &other) = default;
+  constexpr MatrixAddr(ThisType &&other) = default;
+  //constexpr MatrixAddr(const ThisType &other) : offset_(other.offset_) {}
+  //constexpr MatrixAddr(ThisType &&other) : offset_(other.offset_) {}
+
+  ThisType &operator=(const ThisType &) = default;
+  ThisType &operator=(ThisType &&) = default;
+
+  template<typename MatrixAddr__>
+  constexpr MatrixAddr(const MatrixAddr__ &other)
+    : MatrixAddr(other.row(), other.col()) {}
+
+  constexpr uint8_t row() const {
+    return offset_ / cols;
+  }
+  constexpr uint8_t col() const {
+    return offset_ % cols;
+  }
+
+  void setRow(uint8_t r) {
+    //assert(r < rows);
+    offset_ = this->col() + r * cols;
+  }
+  void setCol(uint8_t c) {
+    //assert(c < cols);
+    offset_ = this->row() * cols + c;
+  }
+
+  void shift(int8_t rows, int8_t cols) {
+    offset_ += rows * cols + cols;
+  }
+
+  void rowShift(int8_t rows) {
+    offset_ += rows * cols;
+  }
+
+  void colShift(int8_t cols) {
+    offset_ += cols;
+  }
+
+  constexpr ThisType shifted(int8_t rows, int8_t cols) const {
+    return ThisType(uint8_t(offset_ + rows * cols + cols));
+  }
+
+  constexpr ThisType rowShifted(int8_t rows) const {
+    return ThisType(uint8_t(offset_ + rows * cols));
+  }
+
+  constexpr ThisType colShifted(int8_t cols) const {
+    return ThisType(uint8_t(offset_ + cols));
+  }
+
+  constexpr uint8_t toInt() const {
+    return offset_;
+  }
+
+  constexpr bool isValid() const {
+    return offset_ < upper_limit;
+  }
+
+  ThisType operator++() {
+    ++offset_;
+    return *this;
+  }
+
+  ThisType operator++(int) { // postfix ++
+    ThisType copy(*this);
+    ++*this;         // call the prefix increment
+    return copy;
+  }
+
+  ThisType operator--() {
+    --offset_;
+    return *this;
+  }
+
+  ThisType operator--(int) { // postfix ++
+    ThisType copy(*this);
+    --*this;         // call the prefix increment
+    return copy;
+  }
+
+  template<typename MatrixAddr__>
+  ThisType &operator+(const MatrixAddr__ & other) {
+    *this = ThisType(this->row() + other.row(),
+                     this->col() + other.col());
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  ThisType &operator-(const MatrixAddr__ & other) {
+    *this = ThisType(this->row() - other.row(),
+                     this->col() - other.col());
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  ThisType &operator+=(const MatrixAddr__ & other) {
+    *this = *this + other;
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  ThisType &operator-=(const MatrixAddr__ & other) {
+    *this = *this - other;
+    return *this;
+  }
+
+  bool operator==(const ThisType &other) const {
+    return offset_ == other.offset_;
+  }
+
+  bool operator!=(const ThisType &other) const {
+    return offset_ != other.offset_;
+  }
+
+  const ThisType &operator*() const {
+    return *this;
+  }
+
+  typedef ThisType iterator;
```

Comment:
This is a type name, so it should be `Iterator`, with a leading uppercase character. I think there should also be a paragraph comment here describing its purpose, and how the resulting range-based for loops work.

Reply:
I followed the convention of the STL, but you are right in Kaleidoscope it would be uppercase.